multinet/multinet/multinet.py

```diff
@@ -0,0 +1,290 @@
+"""Flask blueprint for Multinet REST API."""
+import csv
+from graphql import graphql
+from io import StringIO
+import itertools
+import json
+import newick
+import re
+import uuid
+
+from flask import Blueprint, request
+from flask import current_app as app
+
+from .schema import schema
+from . import db
+
+bp = Blueprint('multinet', __name__, url_prefix='/multinet')
+
+
+def graphql_query(query, variables=None):
+    """Perform a GraphQL query using optional variable definitions."""
+    data = None
+    errors = []
+
+    result = graphql(schema, query, variables=variables or {})
+    if result:
+        errors = [error.message for error in result.errors] if result.errors else []
+        app.logger.warn('Errors in request: %s' % len(errors))
+        for error in errors[:10]:
+            app.logger.warn(error)
+    else:
+        data = result.data
+
+    return dict(data=result.data, errors=errors, query=query)
+
+
+def validate_csv(rows):
+    """Perform any necessary CSV validation, and raise appropriate exceptions."""
+    if '_key' in rows.fieldnames:
+        # Node Table, check for key uniqueness
+
+        keys = [row['_key'] for row in rows]
+        uniqueKeys = set()
+        duplicates = set()
+        for key in keys:
+            if key in uniqueKeys:
+                duplicates.add(key)
+            else:
+                uniqueKeys.add(key)
+
+        if (len(duplicates) > 0):
+            return {'error': 'duplicate',
+                    'detail': list(duplicates)}
```

Comment:
I personally don't like how we're returning the error here and handling in the `bulk` method, as it's a bit arbitrary. I think the `bulk` method should be agnostic to the errors thrown here, and whatever is returned here just be minimally wrapped as a tuple and returned from the `bulk` method. However, for viewing errors in detail on the client, we'll need to have a consistent was of returning possibly multiple errors. I think we need to have an offline discussion around this.

Reply:
(Had an offline discussion, summarizing here)

- decided to minimally wrap what is returned from this function
- decided to look closer at returning detailed error objects (whether actual class instances, or formatted dicts/json objects) to describe each error
  - the client can receive the full list and decide what to do with it (just say how many there are; display an error message for each error; etc.)

I will change the code to stop rewrapping the error objects for now, and will open an issue to look more closely at this in the future.

Reply:
See 6694ab5 and #105